### PR TITLE
PKCE Auth Flow

### LINKED ISSUE #150: Authorization Code Flow with Proof Key for Code Exchange (PKCE) is missing
It is a very exciting *Authorization Flow* for developers of server-less applications that use *Spotify API* directly.

Not only will the users of these applications no longer have to setup their own *Spotify Applications*,
but also the creators will get to see usage metrics of their applications in their *Spotify Dashboard*.

You can find out how it works [here](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow-with-proof-key-for-code-exchange-pkce) and [here](https://docs.rs/oauth2/3.0.0/oauth2/#getting-started-authorization-code-grant-w-pkce).

### PR DESCRIPTION
**Pull request by marioortizmanero**
## Description

This implements the PKCE auth flow. It also removes quite a few TODOs in the code.

Closes #150

## Dependencies 

I had to add `sha2` in order to hash the verifier code with SHA256.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

* The `examples/auth_code_pkce.rs` example

### CODE REVIEW

**Comment by marioortizmanero** [comment]
This is ready for review whenever you have time @ramsayleung :)

File: src/clients/mod.rs [hunk 1]
```diff
@@ -84,16 +59,28 @@ mod test {
...
+            Some(&"Bearer access_token".to_owned())
+        );
     }
 
     #[test]
     fn test_basic_auth() {
-        let (auth, value) = basic_auth("ramsay", "123456");
-        assert_eq!(auth, "authorization");
-        assert_eq!(value, "Basic cmFtc2F5OjEyMzQ1Ng==");
+        let creds = Credentials::new_pkce("ramsay");
```
Comment (ramsayleung) [inline src/clients/mod.rs:72]: I think this test should be moved into `lib.rs`, the same file with `struct Credentials`

File: src/clients/mod.rs [hunk 2]
```diff
@@ -84,16 +59,28 @@ mod test {
             ..Default::default()
         };
 
-        let (auth, value) = bearer_auth(&tok);
-        assert_eq!(auth, "authorization");
-        assert_eq!(value, "Bearer access_token");
+        let headers = tok.auth_headers();
```
Comment (ramsayleung) [inline src/clients/mod.rs:62]: Same here, this test should be moved into `rspotify-model/src/auth.rs`, the same file with `struct Token`

**Review by ramsayleung — COMMENTED** [review]
Except for these small nitpicks I commented, everything else looks good to me. Great job!

File: examples/client_creds.rs [hunk 3]
```diff
@@ -5,11 +5,12 @@ async fn main() {
     // You can use any logger for debugging.
     env_logger::init();
 
-    // Set RSPOTIFY_CLIENT_ID, RSPOTIFY_CLIENT_SECRET and
-    // RSPOTIFY_REDIRECT_URI in an .env file or export them manually:
+    // Set RSPOTIFY_CLIENT_ID and RSPOTIFY_CLIENT_SECRET in an .env file or
```
[the remark this answers is not in the document]
Comment (marioortizmanero) [reply examples/client_creds.rs:8]: There's no redirect uri in the client credentials flow

File: src/clients/mod.rs (hunk 1 above)
Comment (marioortizmanero) [reply src/clients/mod.rs:72]: Will do

File: src/clients/mod.rs (hunk 2 above)
Comment (marioortizmanero) [reply src/clients/mod.rs:62]: Yup agreed

**Review by ramsayleung — APPROVED** [review]

**Comment by marioortizmanero** [comment]
Fixed everything, you can merge once the tests pass :)

**Review by ramsayleung — APPROVED** [review]

**Pull request merged**